ipn/store/stores.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package store provides various implementation of ipn.StateStore.
+package store
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"tailscale.com/atomicfile"
+	"tailscale.com/ipn"
+	"tailscale.com/ipn/store/aws"
+	"tailscale.com/ipn/store/kube"
+	"tailscale.com/ipn/store/mem"
+	"tailscale.com/paths"
+	"tailscale.com/types/logger"
+)
+
+// Provider returns a StateStore for the provided path.
+type Provider func(path string) (ipn.StateStore, error)
+
+var regOnce sync.Once
+
+func registerDefaultStores() {
+	Register("mem:", func(string) (ipn.StateStore, error) {
+		return new(mem.Store), nil
+	})
+	Register("file:", func(path string) (ipn.StateStore, error) {
+		path = strings.TrimPrefix(path, "file:")
+		return NewFileStore(path)
+	})
+	Register("kube:", func(path string) (ipn.StateStore, error) {
+		secretName := strings.TrimPrefix(path, "kube:")
+		return kube.New(secretName)
+	})
+	Register("arn:", aws.NewStore)
+}
+
+var knownStores map[string]Provider
+
+// ErrNoMatchingStore is the error used when there is no matching
+// registered store.
+var ErrNoMatchingStore = errors.New("no matching store")
+
+// New returns a StateStore based on the provided path
+// and registered stores.
+// By default the following stores are registered:
+//
+//   * if the string begins with "mem:", the suffix
+//     is ignored and an in-memory store is used.
+//   * if the string begins with "file:", the suffix
```

Comment:
Clarify the relation to NewFileStore here? With the New constructor you need the "file:" prefix but the other one you don't?

Reply:
I made NewFileStore strip out `file:` if provided.